api/src/main/java/org/openmrs/module/aijar/dataintegrity/InvalidTBEncounters.java

```diff
@@ -0,0 +1,210 @@
+		
+		List<Obs> obsList = query.list();
+		Set<Patient> uniquePatientList = new HashSet<Patient>();
+		
+		List<RuleResult<Patient>> ruleResults = new ArrayList<>();
+		for (Obs obs : obsList) {
+			Patient patient = obs.getEncounter().getPatient();
+			RuleResult<Patient> ruleResult = new RuleResult<>();
+			ruleResult.setActionUrl("coreapps/patientdashboard/patientDashboard.page?patientId=" + patient.getUuid());
+			ruleResult.setNotes("Patient #" + getTbNumber(patient, obs.getEncounter()) + " with similar " + identifierTitle + " identifiers(" + obs.getValueText() + "): duplicated for a single patient,");
+			ruleResult.setEntity(patient);
+			
+			ruleResults.add(ruleResult);
+			uniquePatientList.add(patient);
+		}
+		log.info("There are " + uniquePatientList.size() + " Patients with similar " + identifierTitle + " identifiers: duplicated for a single patient,");
+		
+		return ruleResults;
+	}
+	
+	/**
+	 * Patients with missing TB Number identifiers
+	 * @return List<RuleResult<Patient>>
+	 */
+	public List<RuleResult<Patient>> patientsWithMissingTBNumbers(String tbIdentifierConceptUuid, String identifierTitle) {
+		log.info("Executing rule to find patients with missing " + identifierTitle + " TB identifiers");
+
+		String queryString = "SELECT pp.patient FROM PatientProgram pp WHERE pp.voided = 0 AND pp.program.uuid = :tbProgramUuid GROUP BY pp.patient.patientId";
+
+		Query query = getSession().createQuery(queryString);
+		query.setParameter("tbProgramUuid", Programs.TB_PROGRAM.uuid());
+		
+		List<Patient> patientList = query.list();
+		log.info("There are " + patientList.size() + " Patients Enrolled in TB program");
+		
+		List<RuleResult<Patient>> ruleResults = new ArrayList<>();
+		Program tbProgram = Context.getProgramWorkflowService().getProgramByUuid(Programs.TB_PROGRAM.uuid());
+		
+		//For each patient enrolled in the TB program, look for encounters with missing identifier
+		for (Patient patient : patientList) {
+			String encounterQueryString = "FROM Encounter e WHERE e.patient.id = :patientId AND e.encounterType.uuid = :encounterType AND e.id NOT IN (SELECT o.encounter.id FROM Obs o WHERE o.concept.uuid = :conceptUuid)";
+			
+			Query encounterQuery = getSession().createQuery(encounterQueryString);
+			encounterQuery.setParameter("patientId", patient.getId());
+			encounterQuery.setParameter("encounterType", EncounterTypes.TB_SUMMARY_ENCOUNTER.uuid());
+			encounterQuery.setParameter("conceptUuid", tbIdentifierConceptUuid);
+			
+			List<Encounter> patientEncounters = encounterQuery.list();
+			
+			for (Encounter encounter : patientEncounters) {
+				RuleResult<Patient> ruleResult = new RuleResult<>();
+				ruleResult.setActionUrl("htmlformentryui/htmlform/editHtmlFormWithStandardUi.page?patientId=" + patient.getUuid() + "&encounterId=" + encounter.getId());
+				ruleResult.setNotes("Client# " + getTbNumber(patient, encounter) + ", has missing " + identifierTitle + " identifier");
```

Comment:
Change this message to "Patient has a missing XXX identifier" - since at times there are no identifiers so ends up as "Client# , has missing Unit TB Number identifier"